modules/terminate-agent-hook/lambda/lambda_function.py

```diff
@@ -232,26 +325,97 @@ def handler(event, context):
 
-    client = boto3.client("ec2", region_name=event['region'])
+                print(json.dumps({
+                    "Level": "info",
+                    "Message": "Completing lifecycle action..."
+                }))
+                lifecycle_action_response = as_client.complete_lifecycle_action(
+                    AutoScalingGroupName=message['AutoScalingGroupName'],
+                    LifecycleHookName=message['LifecycleHookName'],
+                    LifecycleActionToken=message['LifecycleActionToken'],
+                    LifecycleActionResult="CONTINUE"
+                )
 
+                print(json.dumps({
+                    "Level": "info",
+                    "Message": f"CompleteLifecycleAction Successful, response: {lifecycle_action_response}"
+                }))
+        # catch everything here and log it
+        # pylint: disable=broad-exception-caught
+        except Exception as ex:
+            print(json.dumps({
+                "Level": "exception",
+                "Exception": str(ex)
+            }))
+
+            # if the gitlab-runner service fails to be stopped, the function can error out and the SQS
+            # message will go back to the queue to be retried, up to a set amount of times
+            message_receive_count = int(event['Records'][0]['attributes']['ApproximateReceiveCount'])
+            max_receive_count = int(os.environ['SQS_MAX_RECEIVE_COUNT'])
+
+            print(json.dumps({
+                "Level": "info",
+                "Message": f"Graceful termination retry count: {message_receive_count}/{max_receive_count}"
+            }))
+            if message_receive_count < max_receive_count:
+                print(json.dumps({
+                    "Level": "info",
+                    "Message": "Graceful termination will be retried in next function run"
+                }))
+                sys.exit(1)
+            else:
+                print(json.dumps({
+                    "Level": "info",
+                    "Message": "Reached max received count, continuing with instance termination"
+                }))
+    else:
+        event_detail = event['detail']
+        instance_id = event_detail['EC2InstanceId']
+        region = event['region']
+
+        if event_detail['LifecycleTransition'] != "autoscaling:EC2_INSTANCE_TERMINATING":
+            sys.exit()
+        
+        ec2_client = boto3.client("ec2", region_name=region)
+    
     # make sure that no new instances are created
-    cancel_active_spot_requests(ec2_client=client, executor_name_part=os.environ['NAME_EXECUTOR_INSTANCE'])
+    cancel_active_spot_requests(ec2_client=ec2_client, executor_name_part=os.environ['NAME_EXECUTOR_INSTANCE'])
 
     # find the executors connected to this agent and terminate them as well
-    _terminate_list = ec2_list(client=client, parent=event_detail['EC2InstanceId'])
+    _terminate_list = ec2_list(client=ec2_client, parent=instance_id)
```

Comment:
issue: we shouldn't terminate the instances in case of a graceful shutdown. This is done by the GitLab Runner itself. Otherwise the shutdown is no longer graceful.

Reply:
This section would only run after:
1. graceful terminate is successful
OR
2. we reach the graceful terminate timeout

In the first case, the SSM command should cleanup the workers using `docker-machine rm`, though if for some reason it fails then the workers could be left running.

In the second case, I believe it's possible for the runner instance to be terminated by the lifecycle hook before gitlab runner can remove the workers, leaving the workers orphaned.

I think having this section would act as a back up for these situations.

Reply:
Yes, sounds reasonably to me. Let's keep it here.